src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/EthereumUpstreamValidator.kt

```diff
@@ -68,6 +73,37 @@ open class EthereumUpstreamValidator @JvmOverloads constructor(
         return listOf(results.t1, results.t2, results.t3).sortedWith(cp).last()
     }
 
+    fun validateChainSettings(): Boolean {
+        if (!options.validateChain) {
+            return true
+        }
+        return Mono.zip(
+            chainId(),
+            netVersion()
+        )
+            .map {
+                val chainData = CHAIN_DATA[chain] ?: return@map false
+                val isChainValid = chainData.chainId == it.t1 && chainData.netVersion == it.t2
+
+                if (!isChainValid) {
+                    val actualChain = getChainByData(
+                        DefaultEthereumMethods.HardcodedData(it.t2, it.t1)
+                    )?.chainName
+                    log.warn(
+                        "${chain.chainName} is specified for upstream ${upstream.getId()} " +
+                            "but actually it is $actualChain with chainId ${it.t1} and net_version ${it.t2}"
+                    )
+                }
+
+                isChainValid
+            }
+            .onErrorResume {
+                log.error("Error during chain validation", it)
+                Mono.just(false)
+            }
+            .block() ?: false
```

Comment:
Yea, you are right but this blocking code is in a blocking context. Otherwise we would get an exception. We must do it via blocking code to prevent of creation of such upstream